src/main/java/org/folio/bulkops/service/QueryService.java

```diff
@@ -1,94 +1,318 @@
+
+  private final PermissionsValidator permissionsValidator;
+
+  private final RemoteFileSystemClient remoteFileSystemClient;
+  private final SrsClient srsClient;
+  private final QueryClient queryClient;
+
   private final ExecutorService executor = Executors.newCachedThreadPool();
 
-  public UUID executeQuery(SubmitQuery submitQuery) {
+  public BulkOperation retrieveRecordsAndCheckQueryExecutionStatus(BulkOperation bulkOperation) {
+    executor.execute(getRunnableWithCurrentFolioContext(() -> {
+      var queryId = executeQuery(new SubmitQuery(bulkOperation.getFqlQuery(), bulkOperation.getEntityTypeId()));
+      var queryResult = getResult(queryId);
+      switch (queryResult.getStatus()) {
+          case IN_PROGRESS -> log.info("Retrieving records by FKM for operation {} in progress...", bulkOperation.getId());
+          case SUCCESS -> {
+          if (queryResult.getTotalRecords() == 0) {
+            failBulkOperation(bulkOperation, "No records found for the query");
+          } else {
+            saveIdentifiersAndStartBulkOperation(bulkOperation, queryResult);
+          }
+        }
+        case FAILED -> failBulkOperation(bulkOperation, queryResult.getFailureReason());
+        case CANCELLED -> cancelBulkOperation((bulkOperation));
+      }
+    }));
+    bulkOperation.setStatus(RETRIEVING_RECORDS);
+    bulkOperationRepository.save(bulkOperation);
+    return bulkOperation;
+  }
+
+  private UUID executeQuery(SubmitQuery submitQuery) {
     return queryClient.executeQuery(submitQuery).getQueryId();
   }
 
-  public BulkOperation checkQueryExecutionStatus(BulkOperation bulkOperation) {
-    var queryResult = queryClient.getQuery(bulkOperation.getFqlQueryId());
-    return switch (queryResult.getStatus()) {
-      case SUCCESS -> {
-        if (queryResult.getTotalRecords() == 0) {
-          yield failBulkOperation(bulkOperation, "No records found for the query");
-        }
-        executor.execute(getRunnableWithCurrentFolioContext(() -> saveIdentifiers(bulkOperation)));
-        bulkOperation.setStatus(RETRIEVING_IDENTIFIERS);
-        yield bulkOperationRepository.save(bulkOperation);
-      }
-      case FAILED -> failBulkOperation(bulkOperation, queryResult.getFailureReason());
-      case CANCELLED -> cancelBulkOperation((bulkOperation));
-      case IN_PROGRESS -> bulkOperation;
-    };
+  private QueryDetails getResult(UUID queryId) {
+    return queryClient.getQuery(queryId, true);
   }
 
-  private void saveIdentifiers(BulkOperation bulkOperation) {
+  private void saveIdentifiersAndStartBulkOperation(BulkOperation bulkOperation, QueryDetails queryResult) {
     try {
-      var identifiersString = queryClient.getSortedIds(bulkOperation.getFqlQueryId(), 0, Integer.MAX_VALUE).stream()
-        .map(list -> list.get(0))
-        .distinct()
-        .collect(Collectors.joining(NEW_LINE_SEPARATOR));
+      var identifiers = retrieveIdentifiers(queryResult, getIdField(bulkOperation));
```

Comment:
Saving file with identifiers can be moved to processQueryResult method to populate file together with csv and json files with entities. This case we can avoid double iterating over content array. JIC it is not critical for us to track SAVED_IDENTIFIERS status. It takes quick time even for non-refactored flow. @vashjs , could you please confirm that SAVED_IDENTIFIERS is not used on FE side?